### PR TITLE
[PP] Add eval() API to schedule

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157795

These change add an `eval()` API to PP schedules

## Context

Currently, you can run "Forward only" for a schedule in two ways:
1. Use a custom schedule `_ScheduleForwardOnly`
2. Do not pass in `loss_fn` in schedule constructor, and no backward computations will be executed.

However, this is still limiting because we may want to run forward through the pipeline / calculate the loss, but without backward, e.g. during validation. These changes allow for this.

```python
if self.rank == 0:
    schedule.eval(x)
elif self.rank == self.world_size - 1:
    losses = []
    schedule.eval(target=target, losses=losses)
else:
    schedule.eval()
```

TODO:
- in later PRs, we will deprecate the `_ScheduleForwardOnly`

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

### CODE REVIEW

File: test/distributed/pipelining/test_schedule_multiproc.py [hunk 1]
```diff
@@ -110,6 +110,139 @@ def test_forward_only(self, ScheduleClass):
 
             torch.testing.assert_close(x_clone, out)
 
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @parametrize(
+        "ScheduleClass",
+        [ScheduleGPipe, Schedule1F1B, ScheduleInterleaved1F1B, ScheduleLoopedBFS],
+    )
+    def test_no_grad_context(self, ScheduleClass):
```
Comment (Skylion007) [inline test/distributed/pipelining/test_schedule_multiproc.py]: Should we test torch.inference_mode() context too?

Comment (H-Huang) [reply test/distributed/pipelining/test_schedule_multiproc.py]: that's a good flag, let me add tests for that as well

File: test/distributed/pipelining/test_schedule_multiproc.py [hunk 2]
```diff
@@ -110,6 +110,139 @@ def test_forward_only(self, ScheduleClass):
...
+            ]
+
+            # Test with torch.no_grad() context
+            schedule = ScheduleClass(stages, chunks, loss_fn=loss_fn, scale_grads=False)
+
+            # Clear gradients
+            for stage_module in stage_modules:
+                stage_module.zero_grad()
+
+            with torch.no_grad():
```
Comment (Skylion007) [inline test/distributed/pipelining/test_schedule_multiproc.py]: Just parameterize with `torch.inference_mode()` too

**Comment by wconstab** [comment]
question on the approach:

is your goal to make it so that schedule.step() will behave differently under `no_grad()` and not run backwards?

if so, i wonder if such polymorphic behavior of .step() is really needed.  Should we instead just add an api like `eval()` to the Schedule class, and have this API be equivalent to running ScheduleForwardOnly?
* also, i would be in favor of having this convenience method on all the scheudles and getting rid of ScheduleForwardOnly class

**Comment by H-Huang** [comment]
@wconstab 

> is your goal to make it so that schedule.step() will behave differently under no_grad() and not run backwards?

Yep, that's correct

> Should we instead just add an api like eval() to the Schedule class, and have this API be equivalent to running ScheduleForwardOnly?

This would also work. We could do this instead but the minor API difference is you would need replace all `step()` calls in your code with `eval()`. I figured that we also would want to support `no_grad()` since currently it will raise an exception like "[X] for chunk X has gradients None and is expecting to send gradients to stage X" if run under `no_grad()` so this is fixing two issues.

So I guess we have 3 options:
1) Support step with `torch.no_grad()` and add `eval()` API (they will both functionally do the same thing)
2) Add `eval()` API, add a more user friendly error message in `step()` when using `torch.no_grad()`
3) Only support step with `torch.no_grad()` (current implementation)

Any preferences for these options?

**Comment by wconstab** [comment]
I think my major preference would be to have an explicit eval() api.  Mainly, becuase its clear what it does and we can document it, and also remove the forwardonly schedule.

As for what to do with no_grad and step, i think semantically what 'step' means is "run a train step" and doing a train step with no grad is kind of nonsense.  so i might lean towards raising a nice error there.

File: torch/distributed/pipelining/stage.py [hunk 3]
```diff
@@ -761,6 +760,10 @@ def backward_one_chunk(
         last_backward is controlled by the schedule and signals synchronization of gradients across DP groups
         after the last backward.
         """
+        # skip backward computation if backward is not enabled
+        if not self.has_backward:
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:764]: hm, am i right that the way eval() works is that it calls the regular schedule step() but makes backward steps in the schedule into no-ops?

intuitively i was thinking what you'd do is just have eval call into the 'schedule_forward_only' logic,

however, this approach might be fine too.  Is there some advantage to doing the forward operations in a schedule-specific way?  and is there a risk here that we still accidentally do some extra backward communication because we forgot to disable it, or something?

Comment (H-Huang) [reply torch/distributed/pipelining/stage.py:764]: > am i right that the way eval() works is that it calls the regular schedule step() but makes backward steps in the schedule into no-ops?

Yep that's right.

The advantage of sharing is we can reuse pre and post processing logic we use in step() (microbatch splitting and output merging), and logging that we have in original `step()`. Also because we want to consolidate around the `_PipelineScheduleRuntime` having one method that performs the actual "execution" is easier to reason about I think.

There is a risk as you mentioned of accidentally introducing backward comm, but the hopefully tests should cover this.

Comment (wconstab) [reply torch/distributed/pipelining/stage.py:764]: ok. i guess i'd still _prefer_ if we could keep eval and step separate and just share helpers/utils for microbatching and output merging.  But i agree this is a good reason to use step for now and we could do that later as a cleanup if we want.

**Review by wconstab — APPROVED** [review]
